[Task] Compute MediaType for wireless ifaces in Device Information TLV

mariomaz opened the pull request:
Compute MediaType for wireless ifaces in Device Information TLV of Topology Response message.

mariomaz committed:
tlvf: add value IEEE_802_11AX to enum eMediaType

IEEE 802.11ax is not included in Table 6-12—Media type (intfType) in
the IEEE 1905.1a standard.

WiFi6 is specified to use 0x0108 in R2

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: update autogenerated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bwl: add get_max_bandwidth method to band_info

Add method to compute maximum supported bandwidth by all the channels
in the band.

This value is required to compute MediaType.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: backhaul: complete get_media_type for wireless interfaces

Complete the get_media_type() method to compute MediaType field for
IEEE_802_11 interfaces, according to Table 6-12—Media type (intfType)
found in IEEE 1905.1 standard.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: backhaul: compute MediaType for wireless interfaces in DeviceInformation TLV

Use get_media_type() method to compute the MediaType field for wireless
interfaces included in the DeviceInformation TLV of the Topology
Response message.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
controller: use eMediaTypeGroup to check for wireless interfaces

eMediaTypeGroup is defined as the MSB of the eMediaType, according to
Table 6-12—Media type (intfType) found in IEEE 1905.1a standard.

Instead of using minimum and maximum values of eMediaType to check for
wireless interfaces, use the value of eMediaTypeGroup.

Resulting code is easier to maintain and while we are at it, we also
include the newly created IEEE_802_11AX value.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

rmelotte commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -133,8 +135,55 @@ static bool get_media_type(const std::string &interface_name,
+                beerocks::eFreqType frequency_band     = band_info.get_frequency_band();
+                beerocks::eWiFiBandwidth max_bandwidth = band_info.get_max_bandwidth();
+
+                if ((beerocks::eFreqType::FREQ_24G == frequency_band) &&
```
This block is really hard to read.
Maybe you could first group by frequency, then by bandwidth.
You could also maybe use a switch for the bandwidth
(nitpick)

tomereli replied:
Perhaps this block can be moved to a separate function `max_bandwidth_to_media_type`? (nitpick)

mariomaz replied:
Done

rmelotte reviewed (approved):
:+1:

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eMediaType.h:
```diff
@@ -30,6 +30,7 @@ enum eMediaType: uint16_t {
     IEEE_802_11AC_5_GHZ = 0x105,
     IEEE_802_11AD_60_GHZ = 0x106,
     IEEE_802_11AF = 0x107,
+    IEEE_802_11AX = 0x108,
```
This should be squashed to the previous commit. Generally, when the change is small, we prefer to update the autogenerated files in the same commit as the one which changes the yaml.

mariomaz replied:
Not squashed because the change is not so small after having addressed all the comments in the review.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -133,8 +135,55 @@ static bool get_media_type(const std::string &interface_name,
+         * 8 IEEE 802.11ax (5 GHz)       5 GHz      160 MHz      Not included in Table 6-12—Media type (intfType), WiFi6 is specified to use 0x0108 in R2
+         */
+
+        std::unique_ptr<bwl::nl80211_client> client =
```
I think this is wrong to put this here. The backhaul manager already maintains a radio info map - `std::unordered_map<sMacAddr, sRadioInfo> m_radio_info_map;` which is first filled in `BACKHAUL_ENABLE` message from the son slave, and then on every event - client association, VAPS list update, etc.

A design which will better fit prplmesh architecture is to read all this information in the BWL when the radio info is built, then the backhaul manager will simply have it for every slave.

mariomaz replied:
Done

tomereli reviewed (changes requested):
One major comment - reading radio info should only be done by the BWL backend. We read from hostapd whatever we can, and supplement what's missing with the nl80211_client.
It makes the code easier to maintain and allows us to completely remove the dummy implmentation for the nl80211_client which I think is redundant now that we have the nl80211_client member in the BWL backend base class .

So, I suggest to update `refresh_radio_info` to read the missing values using the nl80211_client and use it eventually in the backhaul manager by parsing the `m_radio_info_map`.

mariomaz committed:
tlvf: add frequency_band and max_bandwidth fields

Add new fields frequency_band and max_bandwidth to struct sNodeHostap
so they can be sent from ap_manager_thread to slave_thread inside
ACTION_APMANAGER_JOINED_NOTIFICATION.

Add those fields to message cACTION_BACKHAUL_ENABLE too so they can be
sent from slave_thread to backhaul_manager.

backhaul_manager stores those fields into m_radio_info_map to later
compute MediaType as per IEEE Std 1905.1, Table 6-12—Media type
(intfType).

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
framework: tlvf: use fully qualified type when computing enum size

Compiling autogenerated code fails if there is a field of an enum type
from a different namespace. Reason is namespace is being trimmed off
when computing enum size.

To fix the problem, call sizeof() with fully qualified type name.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bwl: get frequency band and maximum supported bandwidth using NL80211

Add new fields frequency band and maximum supported bandwidth to
RadioInfo struct.

Use nl80211_client::get_radio_info() to fill in the values for such
new fields when getting radio info.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: send frequency band and max bandwidth from bwl to backhaul manager

Reading radio info should only be done by the BWL backend. We read from
hostapd whatever we can, and supplement what's missing with the
nl80211_client.

Backhaul manager needs frequency band and max bandwidth values to
compute MediaType. These values must be sent from ap_manager_thread to
slave_thread inside cACTION_APMANAGER_JOINED_NOTIFICATION and then to
backhaul_manager inside cACTION_BACKHAUL_ENABLE where are saved into
m_radio_info_map.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>
